balancer-js/src/modules/pools/pool-types/concerns/weighted/join.concern.ts

```diff
@@ -0,0 +1,265 @@
+   * containing the data of the function call on a contract, which can then be sent to the network to be executed.
+   * (ex. [sendTransaction](https://web3js.readthedocs.io/en/v1.2.11/web3-eth.html#sendtransaction)).
+   *
+   * @param {JoinPool}          joinPool - join pool information to be encoded
+   * @param {string}            joinPool.poolId - id of pool being joined
+   * @param {string}            joinPool.sender - account address sending tokens to join pool
+   * @param {string}            joinPool.recipient - account address receiving BPT after joining pool
+   * @param {JoinPoolRequest}   joinPool.joinPoolRequest - object containing join pool request information
+   * @returns {string}          encodedJoinPoolData - Returns an ABI byte string containing the data of the function call on a contract
+   */
+  static encodeJoinPool({
+    poolId,
+    sender,
+    recipient,
+    joinPoolRequest,
+  }: JoinPool): string {
+    const vaultInterface = Vault__factory.createInterface();
+
+    return vaultInterface.encodeFunctionData('joinPool', [
+      poolId,
+      sender,
+      recipient,
+      joinPoolRequest,
+    ]);
+  }
+
+  // Join Concern Interface
+
+  /**
+   * Build join pool transaction parameters with exact tokens in and minimum BPT out based on slippage tolerance
+   * @param {ExactTokensInJoinPoolParameters} params - parameters used to build exact tokens in for bpt out transaction
+   * @param {string}                          params.joiner - Account address joining pool
+   * @param {SubgraphPoolBase}                params.pool - Subgraph pool object of pool being joined
+   * @param {string[]}                        params.tokensIn - Token addresses provided for joining pool (same length and order as amountsIn)
+   * @param {string[]}                        params.amountsIn -  - Token amounts provided for joining pool in EVM amounts
+   * @param {string}                          params.slippage - Maximum slippage tolerance in bps i.e. 50 = 0.5%
+   * @returns                                 transaction request ready to send with signer.sendTransaction
+   */
+  async buildExactTokensInJoinPool({
+    joiner,
+    pool,
+    tokensIn,
+    amountsIn,
+    slippage,
+  }: ExactTokensInJoinPoolParameters): Promise<JoinPoolAttributes> {
+    if (
+      tokensIn.length != amountsIn.length ||
+      tokensIn.length != pool.tokensList.length
+    ) {
+      throw new Error('Must provide amount for all tokens in the pool');
+    }
+
+    const formattedAmountsIn = amountsIn.map((amount, i) => {
+      const relatedToken = pool.tokens.find(
+        (token) => token.address === tokensIn[i]
+      );
+      return formatFixed(amount, relatedToken?.decimals || 18).toString();
+    });
+
+    const [sortedTokensIn, parsedAmountsIn, parsedMinBPTOut] =
+      this.calcBptOutGivenExactTokensIn(
+        pool,
+        tokensIn,
+        formattedAmountsIn,
+        slippage
```

Comment:
I think we should maybe remove slippage from this function and do it explicitly after the BPT amount is calculated.

Reply:
Agree - done ✅